crates/iota-data-ingestion-core/src/reader/v2.rs

```diff
@@ -0,0 +1,717 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    future::Future, num::NonZeroUsize, ops::Range, path::PathBuf, sync::Arc, time::Duration,
+};
+
+use backoff::backoff::Backoff;
+use futures::{StreamExt, future::OptionFuture};
+use iota_config::{
+    node::ArchiveReaderConfig,
+    object_storage_config::{ObjectStoreConfig, ObjectStoreType},
+};
+use iota_metrics::spawn_monitored_task;
+use iota_rest_api::CheckpointData;
+use iota_storage::blob::Blob;
+use iota_types::messages_checkpoint::CheckpointSequenceNumber;
+use object_store::{ObjectStore, path::Path as ObjectStorePath};
+use tokio::{
+    sync::{
+        mpsc::{self, error::TryRecvError},
+        oneshot,
+    },
+    task::JoinHandle,
+    time::timeout,
+};
+use tracing::{debug, error, info};
+
+use crate::{
+    IngestionError, IngestionResult, MAX_CHECKPOINTS_IN_PROGRESS, create_remote_store_client,
+    history::reader::HistoricalReader,
+    reader::{
+        fetch::{
+            CheckpointResult, ReadSource, fetch_from_full_node, fetch_from_object_store,
+            read_local_files_with_retry_and_capacity_check, setup_directory_watcher,
+        },
+        v1::{DataLimiter, gc_processed_files},
+    },
+};
+
+/// Available sources for checkpoint streams supported by the ingestion
+/// framework.
+///
+/// This enum represents the different types of remote sources from which
+/// checkpoint data can be fetched. Each variant corresponds to a supported
+/// backend or combination of backends for checkpoint retrieval.
+pub enum RemoteUrl {
+    /// A REST API endpoint for checkpoint data.
+    Rest(String),
+    /// An object storage backend for checkpoint data.
+    ObjectStore {
+        /// The URL of the object store (e.g., S3 bucket, GCS bucket,
+        /// HTTP/WebDAV Storage).
+        object_store_url: String,
+        /// Additional options for configuring the remote store as key-value
+        /// pairs.
+        remote_store_options: Vec<(String, String)>,
+    },
+    /// A hybrid source combining historical object store and optional live
+    /// object store.
+    HybridHistoricalStore {
+        /// The URL of the historical object store.
+        historical_url: String,
+        /// The URL of the live object store.
+        live_url: Option<String>,
+    },
+}
+
+/// Represents a remote backend for checkpoint data retrieval.
+///
+/// This enum encapsulates the supported remote storage mechanisms that can be
+/// used by the ingestion framework to fetch checkpoint data. Each variant
+/// corresponds to a different type of remote source.
+enum RemoteStore {
+    ObjectStore(Box<dyn ObjectStore>),
```

Comment:
Same as above: we want to deprecate this in the new reader in favor of the hybrid solutions.

Reply:
fixed: https://github.com/iotaledger/iota/pull/7425/commits/00b114b58bbb4aac2d837476f3c95cda6261d61b#diff-9f714a0efbe52313776b7c231fc3b7360c25ea9383f54a95b8e28cb63d0c89d4L79